src/FileAdapter.js

```diff
@@ -1,156 +1,111 @@
-        return this._props
-    }
-
-    get filepath() {
-        return this.props.filepath
-    }
-
-    get dir () {
-        return this.props.dir
-    }
-
-    get path() {
-        return (!this.dir || !this.filepath) ? null : path.resolve(this.dir, this.filepath)
-    }
-
-    get exists() {
-        return this.path && fs.existsSync(path.resolve(this.path))
+class _ {
+  constructor(props) {
+    this._props = { ...props, ...{} }
+  }
+
+  get props() {
+    return this._props
+  }
+
+  get type() {
+    return this._type || _.TYPES.ASSET
+  }
+
+  detectType(fileStream) {
+    if (this._type) {
+      // Not necessary
+      return
     }
 
-    get type () {
-        return this._type || _.TYPES.ASSET
-    }
+    // Figure out the file's extension
+    const stringPathSplit = fileStream.path ? fileStream.path.split('.') : ''
 
-    detectType () {
-        if (this._type) {
-            // Not necessary
-            return 
-        }
+    let fileExtension = stringPathSplit[
+      stringPathSplit.length - 1
+    ].toUpperCase()
 
-        // Figure out the file's extension
-        const ext = path.extname(this.path).toUpperCase().substring(1)
+    fileExtension =
+      fileExtension === 'REMOTE'
+        ? stringPathSplit[stringPathSplit.length - 2].toUpperCase()
+        : fileExtension
 
-        for (let [type, values] of Object.entries(_.TYPES)) {
-            if (values.includes(ext)) {
-                // Looks like we recognize this type
-                this._type = values
-                return 
-            }
-        }
+    for (let [type, values] of Object.entries(_.TYPES)) {
+      if (values.includes(fileExtension)) {
+        // Looks like we recognize this type
+        this._type = type
+        return
+      }
     }
-
-    get isCompilable() {
-        return !_.NONCOMPILABLE_TYPES.includes(this.type)
+  }
+
+  get isCompilable() {
+    return !_.NONCOMPILABLE_TYPES.includes(this.type)
+  }
+
+  process(fileStream) {
+    if (
+      !(fileStream instanceof stream.Stream) ||
+      !fileStream.readable ||
+      !(typeof fileStream._readableState === 'object')
+    ) {
+      return Promise.reject(
+        new Error(_.ERRORS.CANNOT_PROCESS(_.MESSAGES.WRONG_FILE_FORMAT))
+      )
     }
 
-    compile(args, options = {}) {
-        if (!this.isCompilable) {
-            // No need to compile
-            return Promise.resolve()
-        }
+    // assign the file type
+    this.detectType(fileStream)
 
-        return new Promise((resolve, reject) => {
-            try {
-                // Attempt to load the file 
-                const content = fs.readFileSync(this.path, 'utf8')
-
-                if (!content) {
-                    // Next let's make sure we stop right here for empty files
-                    resolve("")
-                    return
-                }
-
-                // Try to parse the file and catch syntax errors
-                const template = ejs.compile(content, {})
```

Comment:
This is good. And you can also do it even if you don't ready from a file. So the file adapter should just give you the stream, and then a CompilableAdapter or something should be a "base" adapter for adapters like JSONAdapter, JSAdapter, MarkdownAdapter, that will extend that and you could have the ```compile``` method in the base one.